src/crimson/os/seastore/seastore_types.h

```diff
@@ -1049,45 +1081,142 @@ class laddr_t {
+    auto base = laddr;
+    base.value += offset >> laddr_t::UNIT_SHIFT;
+    offset &= laddr_t::UNIT_MASK;
+    return loffset_t(base, offset);
+  }
+  static loffset_t minus(const laddr_t &laddr, extent_len_t offset) {
+    auto base = laddr;
+    auto diff = (offset + laddr_t::UNIT_SIZE - 1) >> laddr_t::UNIT_SHIFT;
+    base.value -= diff;
+    offset = (diff << laddr_t::UNIT_SHIFT) - offset;
+    return loffset_t(base, offset);
+  }
+
+  auto operator<=>(const loffset_t&) const = default;
+private:
+  laddr_t base;
+  extent_len_t offset;
+};
+std::ostream &operator<<(std::ostream &out, const loffset_t &loff);
+
+/// =============== operator overload for laddr_t/loffset begin ===============
+inline bool operator==(const laddr_t &laddr, const loffset_t &loff) {
+  return laddr == loff.get_base() && 0 == loff.get_offset();
+}
+inline bool operator==(const loffset_t &loff, const laddr_t &laddr) {
+  return loff.get_base() == laddr && loff.get_offset() == 0;
+}
+
+inline auto operator<=>(const laddr_t &laddr, const loffset_t &loff) {
+  return loffset_t(laddr, 0) <=> loff;
+}
+inline auto operator<=>(const loffset_t &loff, const laddr_t &laddr) {
+  return loff <=> loffset_t(laddr, 0);
+}
+
+inline loffset_t operator+(const laddr_t &laddr, const extent_len_t &off) {
+  return loffset_t::plus(laddr, off);
+}
+inline loffset_t operator+(const loffset_t &loffset, const extent_len_t &length) {
+  return loffset_t::plus(loffset.get_base(), loffset.get_offset() + length);
+}
+inline loffset_t operator+(const extent_len_t &offset, const laddr_t &laddr) {
+  return laddr + offset;
+}
+inline loffset_t operator+(const extent_len_t &length, const loffset_t &loffset) {
+  return loffset + length;
+}
+
+inline laddr_t &operator+=(laddr_t &laddr, extent_len_t len) {
+  laddr = laddr + len;
+  return laddr;
+}
+inline loffset_t &operator+=(loffset_t &loffset, extent_len_t len) {
+  loffset = loffset + len;
+  return loffset;
+}
+
+inline loffset_t operator-(const laddr_t &laddr, const extent_len_t &len) {
+  return loffset_t::minus(laddr, len);
+}
+inline loffset_t operator-(const loffset_t &loffset, const extent_len_t &len) {
+  if (loffset.get_offset() >= len) {
+    return loffset_t(loffset.get_base(), loffset.get_offset() - len);
+  } else {
+    return loffset_t::minus(loffset.get_base(), len - loffset.get_offset());
+  }
+}
+inline laddr_t::Int operator-(const laddr_t &laddr, const loffset_t &loffset) {
+  return laddr - loffset.get_base() - loffset.get_offset();
+}
+inline laddr_t::Int operator-(const loffset_t &loffset, const laddr_t &laddr) {
+  return loffset.get_base() - laddr + loffset.get_offset();
+}
+inline laddr_t::Int operator-(const loffset_t &lhs, const loffset_t &rhs) {
+  return lhs.get_base() - rhs.get_base() + lhs.get_offset() - rhs.get_offset();
+}
+
+inline laddr_t &operator-=(laddr_t &laddr, extent_len_t len) {
+  laddr = laddr - len;
+  return laddr;
+}
+inline loffset_t &operator-=(loffset_t &loffset, extent_len_t len) {
```

Comment:
See other comments, `extent_len_t` should only be used when it is related to the length within an extent.

Let's convert all the incorrect usage to a new type `loffset_t` (replacing the original uint64_t laddr_t), which can accept all the possible unsigned widths related to laddr.

Reply:
Elaborated in https://github.com/ceph/ceph/pull/59182#issuecomment-2295576836